File: aop/aop-annotation-processor/src/main/java/ru/tinkoff/kora/aop/annotation/processor/KoraAspect.java
package ru.tinkoff.kora.aop.annotation.processor;

import com.palantir.javapoet.AnnotationSpec;
import com.palantir.javapoet.CodeBlock;
import com.palantir.javapoet.TypeName;
import com.palantir.javapoet.TypeSpec;

import javax.lang.model.element.ExecutableElement;
import javax.lang.model.type.TypeMirror;
import java.util.ArrayList;
import java.util.List;
import java.util.Set;

public interface KoraAspect {
    Set<String> getSupportedAnnotationTypes();

    interface FieldFactory {
        default String constructorParam(TypeMirror type, List<AnnotationSpec> annotations) {
            return this.constructorParam(TypeName.get(type), annotations);
        }

        String constructorParam(TypeName type, List<AnnotationSpec> annotations);

        default String constructorInitialized(TypeMirror type, CodeBlock initializer) {
            return this.constructorInitialized(TypeName.get(type), initializer);
        }

        String constructorInitialized(TypeName type, CodeBlock initializer);
    }

    sealed interface ApplyResult {
        enum Noop implements ApplyResult {INSTANCE}

        record MethodBody(CodeBlock codeBlock) implements ApplyResult {}
    }

    record AspectContext(TypeSpec.Builder typeBuilder, FieldFactory fieldFactory) {}

    ApplyResult apply(ExecutableElement executableElement, String superCall, AspectContext aspectContext);

    /**
     * Generate calling of original method with params
     *
     * <p>Example result:</p>
     * <pre>{@code super.<originalMethodName>(param1, param2)}</pre>
     *
     * @return codeBlock
     */
    static CodeBlock callSuper(String superCall, List<String> params) {
        var b = CodeBlock.builder()
            .add(superCall)
            .add("(");
        for (int i = 0; i < params.size(); i++) {
            var param = params.get(i);
            if (i > 0) {
                b.add(", ");
            }
            b.add("$N", param);
        }
        b.add(")");
        return b.build();
    }

    /**
     * Generate calling of original method with params
     *
     * <p>Example result:</p>
     * <pre>{@code super.<originalMethodName>(param1, param2)}</pre>
     *
     * @return codeBlock
     */
    static CodeBlock callSuper(ExecutableElement executableElement, String superCall) {
        var params = new ArrayList<String>(executableElement.getParameters().size());
        for (var parameter : executableElement.getParameters()) {
            params.add(parameter.getSimpleName().toString());
        }
        return callSuper(superCall, params);
    }
}

File: aop/aop-symbol-processor/src/main/kotlin/ru/tinkoff/kora/aop/symbol/processor/KoraAspect.kt
package ru.tinkoff.kora.aop.symbol.processor

import com.google.devtools.ksp.symbol.KSFunctionDeclaration
import com.google.devtools.ksp.symbol.KSType
import com.squareup.kotlinpoet.AnnotationSpec
import com.squareup.kotlinpoet.CodeBlock
import com.squareup.kotlinpoet.TypeName
import com.squareup.kotlinpoet.ksp.toTypeName

interface KoraAspect {
    fun getSupportedAnnotationTypes(): Set<String>

    interface FieldFactory {
        fun constructorParam(type: TypeName, annotations: List<AnnotationSpec>): String
        fun constructorParam(type: KSType, annotations: List<AnnotationSpec>): String = constructorParam(type.toTypeName(), annotations)
        fun constructorInitialized(type: TypeName, initializer: CodeBlock): String
        fun constructorInitialized(type: KSType, initializer: CodeBlock): String = constructorInitialized(type.toTypeName(), initializer)
    }

    interface ApplyResult {
        enum class Noop : ApplyResult {
            INSTANCE
        }

        data class MethodBody(val codeBlock: CodeBlock) : ApplyResult
    }

    data class AspectContext(val fieldFactory: FieldFactory)

    fun KSFunctionDeclaration.superCall(superName: String) = superCall(superName, this.parameters.map { it.name?.asString().toString() })

    fun superCall(superName: String, params: Iterable<String>): CodeBlock {
        val b = CodeBlock.builder()
            .add(superName)
            .add("(")
        for ((i, parameter) in params.withIndex()) {
            if (i > 0) {
                b.add(", ")
            }
            b.add("%N", parameter)
        }
        b.add(")")
        return b.build()
    }

    fun apply(ksFunction: KSFunctionDeclaration, superCall: String, aspectContext: AspectContext): ApplyResult
}
